### ISSUE TITLE
Data migrate issue

### ISSUE THREAD

**Issue by rpmilevskiy** [ilyakatz/data-migrate#31]
Hi @ilyakatz 

I have created db migration and data migration with `rails g data_migration AddStatusToUsers`

After I put to db migration
```ruby
add_column :users, :status, :string
```

Also I put to data migration
```ruby
User.update_all(status: 'offline')
```

After I run `rake db:migrate:with_data`

And it run only db migration. It ignored my data migration.

Please could you take a look.

**Comment by ilyakatz** [comment]
what version of the gem are you using?

**Comment by hishammalik** [comment]
Having exactly the same problem. Using version 2.1.0. Got same problem even when I used the gem directly from github. When I do rake db:migrate:status:with_data, I see an messages which suggest that perhaps schema file has not gotten generated.

Running rake data:migrate:status also show the data migration in 'down' state and rake data:migrate does nothing

Rails version is 4.1.9, and ruby version: 2.2.4

**Comment by ilyakatz** [comment]
Hm, that's a problem. Do you have a plain valina rails app or do you do anything special with db connections? 

I'll take a look at in the next few days. In the meantime, can you use previous version 2.0.1

**Comment by ilyakatz** [comment]
@rpmilevskiy @hishammalik I found the problem, it has to do with https://github.com/ilyakatz/data-migrate/issues/22. Basically, the problem is that at some point data and schema migrations were written into `schema_migrations`. I fixed that but in order to be backward compatible I needed to add this piece of code

https://github.com/ilyakatz/data-migrate/blob/master/lib/data_migrate/data_migrator.rb#L28

That look at both `data_migrations` and `schema_migrations` tables. In your case, that presents a problem since both data and schema migrations have the same timestamp. Eg

db/data/20160403190058_test_migration.rb
db/migrate/20160403190058_test_migration.rb

I think a way to fix that is to update data_migrator.rb#L28 to only look at legacy migrations prior to release of the fix (v2.0.1). Any thoughts?

**Comment by hishammalik** [comment]
@ilyakatz my suggestion would be to compromise the backward compatibility and introduce a post-install script for people upgrading from previous versions to do something like `rake data-migration-engine:install` which basically creates the data migration schema table as well as copies all the migrations from schema migration into this table for backward compatibility.

**Comment by ilyakatz** [comment]
@hishammalik that's a good approach for acts-as-taggble and is certainly easier, however, I think the stakes are higher here. It's a good idea to add a readme for the gem, but anecdotally, I can say that it's very easy to miss those for a project with large number of dependencies. 

Problem with this approach is that if someone runs a `rake db:migrate:with_data` in production without upgrade action as per readme they will rerun all previous migrations and can be disastrous.

**Comment by hishammalik** [comment]
@ilyakatz I think you will need to make that compromise to some extent for the long run. It is better to have a separate data_migrations table, and trying to be backward compatible is really not feasible. If you are too concerned about it, then maybe the end resort might be to fork the repo, and create a new gem.

That said, you can additionally check for data_migrations table being empty or not. When `rake db:migrate:with_data` or other rake tasks get run, similar to how they check for the data_migrations table existence, they can check whether, data_migrations table is empty or populated. If its populated, then that is fine, and the task can continue. However, if its not populated, then you can copy the schema_migrations into it (for backward compatibility), give a prompt to user that 'initialization/install' has been performed and then continue with the task at hand.

So basically this technique does an 'auto' install if user forgets to do it themselves. It is backward compatible and somewhat safe too.

**Comment by jfeldstein** [comment]
What if newer versions check for "unmigrated" data migrations in the schema_migrations table, and aborts without running anything if it finds any. 

It could write to Stdout with a message instructing the user to run the installer before running `rake db:migrate:with_data`.

Prevents accidentally migrating data?

**Comment by jfeldstein** [comment]
Fwiw, I'm using `2.1.0`. 

When I ran `rails g data_migration`, it gave me two similarly numbered migrations. When I run `rake db:migrate:with_data` it only runs the schema migration. 

Did I forget some part of the setup?

**Comment by bilby91** [comment]
Running the same issue here :(. Any update on a fix ?

**Comment by hishammalik** [comment]
@ilyakatz a simple fix would be to generate a timestamp with slight variation on the timestamp between data and db migrations. Change can be a difference of a second. This would provide backward compatibility.

**Comment by ilyakatz** [comment]
Apologies folks, i've been busy/away and lost track of this. @hishammalik this is a great idea, and easily implemented indeed. Let me try to do this today

**Comment by ilyakatz** [comment]
So as I was working on this, I thought of something. If we automatically create `data` migration and `database` migration, like this

`db/data/20160726215606_something_something.rb`
`db/migrate/20160726215606_something_something.rb`

It's not entirely intuitive which one should run first (data migration or schema migration). And this can be a big deal at times. If we add a millisecond as @hishammalik proposed, that will make it more obvious. However, I'm starting to doubt the idea that both of these migrations should be created at the same time. After all, rails already has an easy way to creation migrations, so why not just stick to that for schema. I am thinking of removing the implicit option. Any strong objections?

**Comment by costajohnt** [comment]
I agree that rails makes it easy to create schema migrations.  Maybe the ability to do both at the same time is useful to some, but I always skip the schema migration.  This is a data migration gem, so maybe keeping it simple and only having it migrate data is the way to go.  👍

**Comment by DulceDeMochi** [comment]
Every time I've needed to do a data migration, I’ve just deleted the DB migration as I didn’t need it. I could see needing to do one then the other, but I’d rather make that decision myself.

**Comment by joegesualdo** [comment]
+1 for smaller, more focused gems.

**Comment by hishammalik** [comment]
Yes, I agree. most of the times, I'm just interested in data migration and have to delete the db migration manually as I don't remember the flag to generate db migration.

**Comment by VagishVela** [comment]
@ilyakatz Are we planning to remove db migrations from this gem? If you need some assistance I can probably remove it and create a PR for it.

**Comment by ilyakatz** [comment]
Thanks for the reminder. I actually had a branch that I stared working on, could you test it out if you have  chance 

```ruby
gem "data_migrate", github: "ilyakatz/data-migrate", branch: "remove_db_migrate"
```

**Comment by VagishVela** [comment]
@ilyakatz I've tested it out in a couple of different projects and it works fine!

**Comment by ilyakatz** [comment]
thanks for checking, @VagishVela new version released to rubygems
